src/Lean/Meta/Tactic/FunInd.lean

```diff
@@ -0,0 +1,855 @@
+
+For a non-mutual, unary function `foo` (or else for the `_unary` function), we
+
+1. expect its definition, possibly after some `whnf`’ing, to be of the form
+   ```
+   def foo := fun x₁ … xₙ (y : a) => fix (fun y' oldIH => body) y
+   ```
+   where `xᵢ…` are the fixed prefix and `y` is parameter of the function.
+
+2. From this structure we derive the type of the `motive`, and start assembling the induction
+   principle:
+   ```
+   def foo.induct := fun x₁ … xₙ (motive : (y : a) → Prop) =>
+    fix (fun y' newIH => T[body])
+   ```
+
+3. The first phase, transformation `T1[body]` (implemented in) `buildInductionBody`,
+   mirrors the branching structure of `foo`, i.e. replicates `ite` and matcher applications,
+   while adjusting their motive. It also unfolds call to `oldIH` and collects induction hypotheses
+   in conditions (see below).
+
+   In particular when translating a `match`, it is prepared to see the idiom
+   as introduced by `mkFix` via `Lean.Meta.MatcherApp.addArg?`, which refines the type of `oldIH`
+   throught the match. The transformation will replace `oldIH` with `newIH` here.
+   ```
+        T[(match (motive := fun oldIH => …) y with | … => fun oldIH' => body) oldIH]
+    ==> (match (motive := fun newIH => …) y with | … => fun newIH' => T[body]) newIH
+   ```
+
+4. When a tail position (no more branching) is found, function `buildInductionCase` assembles the
+   type of the case: A fresh `MVar` asserts the current goal, unwanted values from the local context
+   are cleared, and the current `body` is frisked for induction hypotheses using `collectIHs`,
+   which are then asserted in the `MVar`.
+
+
+5. Function `collectIHs` walks the term and collects the induction hypotheses for the current case
+   (with proofs). When it encounters a saturated application of `oldIH x proof`, it returns
+   `newIH x proof : motive x`.
+
+   Since `x` and `proof` can contain further recursive calls, it uses
+   `foldCalls` to replace these with calls to `foo`. This assumes that the
+   termination proof `proof` works nevertheless.
+
+   Again, we may encounter the `Lean.Meta.Matcherapp.addArg?` idiom, and again we thread `newIH`
+   through, replacing the extra argument. The resulting type of this induction hypothesis is now
+   itself a `match` statement (cf. `Lean.Meta.MatcherApp.inferMatchType`)
+
+   The termination proof of `foo` may have abstracted over some proofs; we need to transfer these
+   proofs, so we unfold these auxillary lemmas if needed.
+
+6. The function `foldCalls` replaces calls to `oldIH` with calls to `foo` that
+   make sense to the user.
+
+   At the end of this transformation, no mention of `oldIH` must remain.
+
+7. After this construction, the MVars introduced by `buildInductionCase` are turned into parameters.
+
+The resulting term then becomes `foo.induct` at its inferred type.
+
+If `foo` is not unary and/or part of a mutual reduction, then the induction theorem for `foo._unary`
```

Comment:
Readers may not know about `_unary` - the first mention of it should briefly summarize what it is.